File: solutions/76-minimum-window-substring/minimum-window-substring.go
func minWindow(s string, t string) string {
    if len(s) == 0 || len(t) == 0 {
        return ""
    }
    dictT := make(map[rune]int)
    for _, char := range t {
        dictT[char]++
    }
    required := len(dictT)
    filteredS := []struct {
        index   int
        element rune
    }{}
    for i, char := range s {
        if _, ok := dictT[char]; ok {
            filteredS = append(filteredS, struct {
                index   int
                element rune
            }{index: i, element: char})
        }
    }
    l, r, formed := 0, 0, 0
    windowCounts := make(map[rune]int)
    ans := []int{-1, 0, 0}
    for r < len(filteredS) {
        char := filteredS[r].element
        windowCounts[char]++
        if count, ok := dictT[char]; ok && windowCounts[char] == count {
            formed++
        }
        for l <= r && formed == required {
            char := filteredS[l].element
            end := filteredS[r].index
            start := filteredS[l].index
            if ans[0] == -1 || end-start+1 < ans[0] {
                ans[0] = end - start + 1
                ans[1] = start
                ans[2] = end
            }
            windowCounts[char]--
            if count, ok := dictT[char]; ok && windowCounts[char] < count {
                formed--
            }
            l++
        }
        r++
    }
    if ans[0] == -1 {
        return ""
    } else {
        return s[ans[1] : ans[2]+1]
    }
}

File: solutions/76-minimum-window-substring/minimum-window-substring.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        struct element {
            int index;
            char char_val;
        };
        if (s.length() == 0 || t.length() == 0) {
            return "";
        }
        unordered_map<char, int> dictT;
        for (int i = 0; i < t.length(); i++) {
            dictT[t[i]]++;
        }
        int required = dictT.size();
        vector<element> filteredS;
        for (int i = 0; i < s.length(); i++) {
            if (dictT.find(s[i]) != dictT.end()) filteredS.push_back({i, s[i]});
        }
        int l = 0, r = 0, formed = 0;
        unordered_map<char, int> windowCounts;
        vector<int> ans = {-1, 0, 0};
        while (r < filteredS.size()) {
            char c = filteredS[r].char_val;
            windowCounts[c]++;
            if (dictT.find(c) != dictT.end() && windowCounts[c] == dictT[c]) {
                formed++;
            }
            while (l <= r && formed == required) {
                char c = filteredS[l].char_val;
                int end = filteredS[r].index;
                int start = filteredS[l].index;
                if (ans[0] == -1 || end - start + 1 < ans[0]) {
                    ans[0] = end - start + 1;
                    ans[1] = start;
                    ans[2] = end;
                }
                windowCounts[c]--;
                if (dictT.find(c) != dictT.end() &&
                    windowCounts[c] < dictT[c]) {
                    formed--;
                }
                l++;
            }
            r++;
        }
        return ans[0] == -1 ? "" : s.substr(ans[1], ans[0]);
    }
};